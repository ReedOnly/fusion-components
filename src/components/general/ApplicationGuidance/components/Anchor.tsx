import React from 'react';
import {
    AnchorDOMRect,
    OverlayAnchorElement,
    OverlayAnchorElementTag,
    OverlayAnchorElementProps,
    OverlayAnchorConnectEvent,
} from '../../../../customElements/components/overlay/anchor';

export interface AppGuideAnchorRef<R extends HTMLElement> {
    id: string;
    scope?: string;
    padding?: number;
    ref: React.RefObject<R>;
}

export const useAnchor = <R extends HTMLElement>(anchor: Omit<AppGuideAnchorRef<R>, 'ref'>) => {
    const ref = React.useRef<R>(null);
    useAnchorRef({ ...anchor, ref });
    return ref;
};

export const useAnchorRef = <R extends HTMLElement>(anchor: AppGuideAnchorRef<R>) => {
    const { id, ref, scope, padding } = anchor;
    const callBackRef = React.useRef<VoidFunction>();
    const anchorPadding = React.useRef<number>(padding);

    React.useEffect(() => {
        if (!ref.current) {
            return;
        }
        requestAnimationFrame(() => {
            const event = new OverlayAnchorConnectEvent({
                detail: {
                    id,
                    scope,
                    bounds: () => {
                        return AnchorDOMRect.create(
                            ref.current.getBoundingClientRect(),
                            anchorPadding.current
                        );
                    },
                    disconnectedCallback: (cb: VoidFunction) => {
                        callBackRef.current = cb;
                    },
                },
                cancelable: false,
                bubbles: true,
                composed: true,
            });
            ref.current.dispatchEvent(event);
        });
        return () => callBackRef.current();
    }, [ref]);
};

export type ApplicationGuidanceAnchorProps = React.PropsWithChildren<
    Omit<OverlayAnchorElementProps, 'bounds'> &
        React.DetailedHTMLProps<React.HTMLAttributes<OverlayAnchorElement>, OverlayAnchorElement>
>;

export const ApplicationGuidanceAnchor: React.FC<ApplicationGuidanceAnchorProps> = (
    props: ApplicationGuidanceAnchorProps
) => {
    return <fusion-overlay-anchor {...props}></fusion-overlay-anchor>;
};

declare global {
    namespace JSX {
        interface ReactHTML {
            [OverlayAnchorElementTag]: React.DetailedHTMLFactory<
                ApplicationGuidanceAnchorProps,
                OverlayAnchorElement
            >;
        }
        interface IntrinsicElements {
            [OverlayAnchorElementTag]: React.DetailedHTMLProps<
                ApplicationGuidanceAnchorProps,
                OverlayAnchorElement
            >;
        }
    }
}

export default ApplicationGuidanceAnchor;
