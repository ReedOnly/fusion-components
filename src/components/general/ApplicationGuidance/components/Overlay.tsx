import React from 'react';
import {
    OverlayElement,
    OverlayElementProps,
    OverlayElementTag,
    OverlayEvent,
    OverlayEventType,
} from '../../../../customElements/components/overlay';

export type OverlayProps = React.PropsWithChildren<
    OverlayElementProps &
        React.DetailedHTMLProps<React.HTMLAttributes<OverlayElement>, OverlayElement> & {
            onSelection?: (e: OverlayEvent<OverlayEventType.selection>) => void;
            onScope?: (e: OverlayEvent<OverlayEventType.scope>) => void;
        }
>;

declare global {
    namespace JSX {
        interface ReactHTML {
            [OverlayElementTag]: React.DetailedHTMLFactory<OverlayProps, OverlayElement>;
        }
        interface IntrinsicElements {
            [OverlayElementTag]: React.DetailedHTMLProps<OverlayProps, OverlayElement>;
        }
    }
}

export const Overlay: React.FC<OverlayProps> = (args: OverlayProps) => {
    const { scope, onSelection, onScope, ...props } = args;
    const ref = React.useRef<OverlayElement>();
    React.useEffect(() => {
        if (!ref.current) return;
        const el = ref.current;
        el.scope = scope;
        onSelection && el.addEventListener(OverlayEventType.selection, onSelection);
        onScope && el.addEventListener(OverlayEventType.scope, onScope);
        return () => {
            onSelection && el.removeEventListener(OverlayEventType.selection, onSelection);
            onScope && el.removeEventListener(OverlayEventType.scope, onScope);
        };
    }, [ref]);

    React.useEffect(() => {
        if (!ref.current) return;
        ref.current.scope = scope;
    }, [ref, scope]);

    return <fusion-overlay ref={ref} {...props}></fusion-overlay>;
};

export default Overlay;
