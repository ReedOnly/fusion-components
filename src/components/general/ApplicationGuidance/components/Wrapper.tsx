import React from 'react';
import {
    ApplicationGuideElement,
    ApplicationGuideElementProps,
    ApplicationGuideElementTag,
    ApplicationGuideEvent,
    ApplicationGuideEventType,
} from '../../../../customElements/components/application-guide';

export type ApplicationGuidanceWrapperProps = React.PropsWithChildren<
    ApplicationGuideElementProps &
        React.DetailedHTMLProps<
            React.HTMLAttributes<ApplicationGuideElement>,
            ApplicationGuideElement
        > & {
            onOpen?: (e: ApplicationGuideEvent<ApplicationGuideEventType.activated>) => void;
            onClose?: (e: ApplicationGuideEvent<ApplicationGuideEventType.deactivated>) => void;
            onShow?: (e: ApplicationGuideEvent<ApplicationGuideEventType.show>) => void;
        }
>;

declare global {
    namespace JSX {
        interface ReactHTML {
            [ApplicationGuideElementTag]: React.DetailedHTMLFactory<
                ApplicationGuidanceWrapperProps,
                ApplicationGuideElement
            >;
        }
        interface IntrinsicElements {
            [ApplicationGuideElementTag]: React.DetailedHTMLProps<
                ApplicationGuidanceWrapperProps,
                ApplicationGuideElement
            >;
        }
    }
}

export const ApplicationGuidanceWrapper: React.FC<ApplicationGuidanceWrapperProps> = (
    args: ApplicationGuidanceWrapperProps
) => {
    const { onOpen, onClose, onShow, ...props } = args;
    const ref = React.useRef<ApplicationGuideElement>();
    React.useEffect(() => {
        if (!ref.current) return;
        const el = ref.current;
        onOpen && el.addEventListener(ApplicationGuideEventType.activated, onOpen);
        onClose && el.addEventListener(ApplicationGuideEventType.deactivated, onClose);
        onShow && el.addEventListener(ApplicationGuideEventType.show, onShow);
        return () => {
            onOpen && el.removeEventListener(ApplicationGuideEventType.activated, onOpen);
            onClose && el.removeEventListener(ApplicationGuideEventType.deactivated, onClose);
            onShow && el.removeEventListener(ApplicationGuideEventType.show, onShow);
        };
    }, [ref]);
    return <fusion-application-guide ref={ref} {...props}></fusion-application-guide>;
};

export default ApplicationGuidanceWrapper;
