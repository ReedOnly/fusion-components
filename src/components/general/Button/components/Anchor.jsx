import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const AnchorComponent = forwardRef(
    (
        {
            children,
            disabled,
            url,
            targetBlank,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <a
            className={getButtonClasses(props)}
            disabled={disabled}
            href={url}
            target={targetBlank ? "_blank" : "_self"}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </a>
    )
);

AnchorComponent.propTypes = {
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
    disabled: PropTypes.bool,
    url: PropTypes.string.isRequired,
    targetBlank: PropTypes.bool,
    onClick: PropTypes.func,
    onClickCapture: PropTypes.func,
    onMouseDown: PropTypes.func,
    onMouseUp: PropTypes.func,
};

AnchorComponent.defaultProps = {
    disabled: true,
    targetBlank: false,
    onClick: null,
    onClickCapture: null,
    onMouseDown: null,
    onMouseUp: null,
};

AnchorComponent.displayName = "@fusion/components/general/Button/AnchorComponent";

export default AnchorComponent;
