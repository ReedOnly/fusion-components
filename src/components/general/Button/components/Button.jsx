import React, { forwardRef } from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const ButtonComponent = forwardRef(
    (
        {
            children,
            disabled,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.displayName =
    "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
