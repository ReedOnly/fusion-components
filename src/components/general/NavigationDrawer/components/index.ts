export { default as MenuChild } from "./Child";