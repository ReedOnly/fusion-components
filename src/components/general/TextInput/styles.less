@import '~styles/mixins.less';

.inputContainer{
    display: flex;
    flex-direction: column;

    
    .inputContent {
        display:  flex;
        justify-content: space-between;
        min-height: .grid-unit(5px)[];
        max-height: .grid-unit(7px)[];
        max-width: .grid-unit(41px)[];
        background-color: var(--color-black-alt5);
        border-bottom: 1px solid var(--color-primary-accent);
        caret-color: var(--color-primary);
        padding: 0 .grid-unit(2px)[];

        &:hover {
            cursor: text;
        }

        &.disabled {
          border-bottom-color: var(--color-black-alt5);
          cursor: not-allowed;

          .inputTextContent{
            .label {
                color: var(--color-primary-accent);
                opacity: 0.6;
            }
            input{
                cursor: not-allowed;
            }
        }
        }
        &.focus {
            border-bottom: 2px solid var(--color-primary);
    
            &.error{
                border-bottom-color: var(--color-red);
            } 
            .inputTextContent{
                padding-bottom: calc(.grid-unit(1px)[] - 1px);

            }
        }
    
        &.error {
            border-bottom-color: var(--color-red);
            caret-color: var(--color-red);

            .inputTextContent{
                .label {
                    color: var(--color-red);
                }
            }
        }

        .inputTextContent{
            display: flex;
            flex-direction: column;
            padding-bottom: .grid-unit(1px)[];
            width: 100%;
            
            &.labelLess{
                padding-top: .grid-unit(1px)[];
            }

            .label {
                font-size: 12px;
                line-height: 12px;
                padding-top: .grid-unit(1px)[];
            }

            input {
                font-size: 16px;
                line-height: .grid-unit(3px)[];
                width: 100%;
                outline: none;
                background-color: var(--color-black-alt5);
                border:none;
     
                &::placeholder {
                    color: var(--color-primary-accent);
                    opacity: 0.6;
                }
            }
        }
        .icon {
            align-self: center;
            padding-left: .grid-unit(2px)[];

            svg{
                width: .grid-unit(3px)[];
                height: .grid-unit(3px)[];
            }
        }
    }
    .helperText{
        padding: 0 .grid-unit(2px)[];
        font-size: 11px;
        line-height: .grid-unit(2px)[];
        color: var(--color-black-alt1);

        &.error {
            color: var(--color-red);
        }
    }
}