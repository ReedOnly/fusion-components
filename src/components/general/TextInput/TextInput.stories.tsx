import * as React from 'react';
import { storiesOf } from '@storybook/react';
import TextInput from './index';
import { SearchIcon } from '../../icons/components/action';
import Spinner from '../../feedback/Spinner';
import useKeyboardNavigation from '../../../hooks/useKeyboardNavigation';

const TextInputStory = () => {
    const [value, setValue] = React.useState('');
    const [loading, setLoading] = React.useState(false);
    const [hasError, setHasError] = React.useState(false);
    const inputRef = React.useRef<HTMLInputElement | null>(null);

    const simulateLoad = () => {
        setLoading(true);
        setTimeout(() => {
            setLoading(false);
            value.length <= 8 && setHasError(true);
        }, 3000);
    };
    React.useEffect(() => {
        if (hasError) {
            setHasError(false);
        }
    }, [value]);

    useKeyboardNavigation(
        {
            onEnter: () => simulateLoad(),
        },
        inputRef.current
    );

    return (
        <React.Fragment>
            <TextInput
                onChange={value => setValue(value)}
                value={value}
                label="Text Input"
                isOptional
                helperText="Input text"
            />
            <br />
            <TextInput
                error
                label="Error input"
                errorMessage="An error occurred"
                onChange={value => setValue(value)}
                value={value}
            />
            <br />
            <TextInput
                onChange={value => setValue(value)}
                value={value}
                disabled
                label="Disabled"
            />
            <br />
            <TextInput
                onChange={value => setValue(value)}
                value={value}
                icon={!loading ? <SearchIcon color="#666666" /> : <Spinner inline />}
                onIconAction={simulateLoad}
                error={hasError}
                errorMessage="Error: At least 8 characters"
                label="Search"
                ref={inputRef}
            />
            <br />
        </React.Fragment>
    );
};

storiesOf('General|TextInput', module).add('Default', () => <TextInputStory />);
