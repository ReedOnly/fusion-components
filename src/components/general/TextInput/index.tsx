import * as React from 'react';
import styles from './styles.less';
import classNames from 'classnames';

type TextInputProps = {
    disabled?: boolean;
    error?: boolean;
    helperText?: string;
    isRequired?: boolean;
    placeholder?: string;
    label?: string;
    onChange: (newValue: string) => void;
    value?: string;
    icon?: HTMLOrSVGImageElement;
    onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
};

const TextInput: React.FC<TextInputProps> = ({
    disabled = false,
    error,
    helperText,
    isRequired = false,
    placeholder = '',
    label,
    onChange,
    value = '',
    icon,
    onBlur,
    ...rest
}) => {
    const inputRef = React.useRef<HTMLInputElement>(null);
    const [focus, setFocus] = React.useState(false);

    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (!disabled) {
            const newValue = event.target.value;
            onChange(newValue);
        }
    };

    const setInputFocus = () => {
        if (inputRef.current) {
            inputRef.current.focus();
            setFocus(true);
        }
    };
    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        if (!disabled) {
            setFocus(false);
            onBlur && onBlur(event);
        }
    };

    const inputContainerClasses = classNames(styles.inputContainer, {
        [styles.focus]: focus,
        [styles.error]: error,
    });

    return (
        <div className={inputContainerClasses} onClick={setInputFocus}>
            <span className={styles.label}>{label}</span>
            <div className={styles.inputContent}>
                <input
                    ref={inputRef}
                    placeholder={placeholder}
                    onBlur={handleBlur}
                    value={value}
                    onChange={handleChange}
                    {...rest}
                />
                <div className={styles.icon}>{icon}</div>
            </div>
        </div>
    );
};

export default TextInput;
