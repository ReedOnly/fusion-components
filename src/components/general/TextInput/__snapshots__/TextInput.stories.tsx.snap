// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|TextInput Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "padding": "8px",
        }
      }
    >
      <div
        className="inputContainer"
      >
        <div
          className="inputContent"
          onClick={[Function]}
        >
          <div
            className="inputTextContent"
          >
            <label>
              Label text
            </label>
            <input
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              value=""
            />
          </div>
        </div>
        <div
          className="helperText"
        >
          Helper text - Optional
        </div>
      </div>
      <br />
      <div
        className="inputContainer"
      >
        <div
          className="inputContent error"
          onClick={[Function]}
        >
          <div
            className="inputTextContent error"
          >
            <label>
              Error text
            </label>
            <input
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              value=""
            />
          </div>
          <div
            className="icon"
          >
            <svg
              height={24}
              style={
                Object {
                  "color": "var(--color-red)",
                }
              }
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                clipRule="evenodd"
                d="M12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2ZM12 13C11.45 13 11 12.55 11 12V8C11 7.45 11.45 7 12 7C12.55 7 13 7.45 13 8V12C13 12.55 12.55 13 12 13ZM11 15V17H13V15H11Z"
                fill="currentColor"
                fillRule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          className="helperText error"
        >
          An error occurred
        </div>
      </div>
      <br />
      <div
        className="inputContainer"
      >
        <div
          className="inputContent disabled"
          onClick={[Function]}
        >
          <div
            className="inputTextContent disabled"
          >
            <label>
              Disabled
            </label>
            <input
              disabled={true}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              value=""
            />
          </div>
        </div>
      </div>
      <br />
      <div
        className="inputContainer"
      >
        <div
          className="inputContent labelLess"
          onClick={[Function]}
        >
          <div
            className="inputTextContent"
          >
            <input
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              value=""
            />
          </div>
          <div
            className="icon"
            onClick={[Function]}
          >
            <svg
              cursor="pointer"
              height={24}
              style={
                Object {
                  "color": "#666666",
                }
              }
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                clipRule="evenodd"
                d="M15.1866 14.4716H15.9766L20.2166 18.7316C20.6266 19.1416 20.6266 19.8116 20.2166 20.2216C19.8066 20.6316 19.1366 20.6316 18.7266 20.2216L14.4766 15.9716V15.1816L14.2066 14.9016C12.8066 16.1016 10.8966 16.7216 8.86658 16.3816C6.08658 15.9116 3.86658 13.5916 3.52658 10.7916C3.00658 6.56156 6.56658 3.00156 10.7966 3.52156C13.5966 3.86156 15.9166 6.08156 16.3866 8.86156C16.7266 10.8916 16.1066 12.8016 14.9066 14.2016L15.1866 14.4716ZM5.47658 9.97156C5.47658 12.4616 7.48658 14.4716 9.97658 14.4716C12.4666 14.4716 14.4766 12.4616 14.4766 9.97156C14.4766 7.48156 12.4666 5.47156 9.97658 5.47156C7.48658 5.47156 5.47658 7.48156 5.47658 9.97156Z"
                fill="currentColor"
                fillRule="evenodd"
              />
            </svg>
          </div>
        </div>
      </div>
      <br />
      <div
        className="inputContainer"
      >
        <div
          className="inputContent"
          onClick={[Function]}
        >
          <div
            className="inputTextContent"
          >
            <label>
              Masked text input
            </label>
            <input
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
